### ISSUE TITLE
Validation fails on HTML in foreign object

### ISSUE THREAD

**Issue by KDean-Dolphin** [validator/validator#1816]
The SVG [foreignObject element](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Element/foreignObject) "includes elements from a different XML namespace. In the context of a browser, it is most likely (X)HTML."

The [br](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-br-element) and [wbr](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-wbr-element) elements are explicitly documented as having no end tag. The JavaScript DOM enforces this; if a document is constructed with `<br />` or `<br></br>`, it will be rewritten during processing as `<br>`.

This happens a lot in Mermaid, which generates SVG containing HTML for textual content; line breaks in the text are rendered internally with `<br />`, but the final output is `<br>`, which fails validation.

Here's a sample SVG:

```html
<svg id="diagram" width="100%" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" class="flowchart" style="max-width: 272.703125px; transform: translate(0px) scale(1);" viewBox="-48 -48 272.703125 286" aria-roledescription="flowchart-v2" height="286" preserveAspectRatio="xMinYMin">
  <style>#diagram{font-family:"trebuchet ms",verdana,arial,sans-serif;font-size:16px;fill:#ccc;}#diagram .error-icon{fill:#a44141;}#diagram .error-text{fill:#ddd;stroke:#ddd;}#diagram .edge-thickness-normal{stroke-width:1px;}#diagram .edge-thickness-thick{stroke-width:3.5px;}#diagram .edge-pattern-solid{stroke-dasharray:0;}#diagram .edge-thickness-invisible{stroke-width:0;fill:none;}#diagram .edge-pattern-dashed{stroke-dasharray:3;}#diagram .edge-pattern-dotted{stroke-dasharray:2;}#diagram .marker{fill:lightgrey;stroke:lightgrey;}#diagram .marker.cross{stroke:lightgrey;}#diagram svg{font-family:"trebuchet ms",verdana,arial,sans-serif;font-size:16px;}#diagram p{margin:0;}#diagram .label{font-family:"trebuchet ms",verdana,arial,sans-serif;color:#ccc;}#diagram .cluster-label text{fill:#F9FFFE;}#diagram .cluster-label span{color:#F9FFFE;}#diagram .cluster-label span p{background-color:transparent;}#diagram .label text,#diagram span{fill:#ccc;color:#ccc;}#diagram .node rect,#diagram .node circle,#diagram .node ellipse,#diagram .node polygon,#diagram .node path{fill:#1f2020;stroke:#ccc;stroke-width:1px;}#diagram .rough-node .label text,#diagram .node .label text,#diagram .image-shape .label,#diagram .icon-shape .label{text-anchor:middle;}#diagram .node .katex path{fill:#000;stroke:#000;stroke-width:1px;}#diagram .rough-node .label,#diagram .node .label,#diagram .image-shape .label,#diagram .icon-shape .label{text-align:center;}#diagram .node.clickable{cursor:pointer;}#diagram .root .anchor path{fill:lightgrey!important;stroke-width:0;stroke:lightgrey;}#diagram .arrowheadPath{fill:lightgrey;}#diagram .edgePath .path{stroke:lightgrey;stroke-width:2.0px;}#diagram .flowchart-link{stroke:lightgrey;fill:none;}#diagram .edgeLabel{background-color:hsl(0, 0%, 34.4117647059%);text-align:center;}#diagram .edgeLabel p{background-color:hsl(0, 0%, 34.4117647059%);}#diagram .edgeLabel rect{opacity:0.5;background-color:hsl(0, 0%, 34.4117647059%);fill:hsl(0, 0%, 34.4117647059%);}#diagram .labelBkg{background-color:rgba(87.75, 87.75, 87.75, 0.5);}#diagram .cluster rect{fill:hsl(180, 1.5873015873%, 28.3529411765%);stroke:rgba(255, 255, 255, 0.25);stroke-width:1px;}#diagram .cluster text{fill:#F9FFFE;}#diagram .cluster span{color:#F9FFFE;}#diagram div.mermaidTooltip{position:absolute;text-align:center;max-width:200px;padding:2px;font-family:"trebuchet ms",verdana,arial,sans-serif;font-size:12px;background:hsl(20, 1.5873015873%, 12.3529411765%);border:1px solid rgba(255, 255, 255, 0.25);border-radius:2px;pointer-events:none;z-index:100;}#diagram .flowchartTitleText{text-anchor:middle;font-size:18px;fill:#ccc;}#diagram rect.text{fill:none;stroke-width:0;}#diagram .icon-shape,#diagram .image-shape{background-color:hsl(0, 0%, 34.4117647059%);text-align:center;}#diagram .icon-shape p,#diagram .image-shape p{background-color:hsl(0, 0%, 34.4117647059%);padding:2px;}#diagram .icon-shape rect,#diagram .image-shape rect{opacity:0.5;background-color:hsl(0, 0%, 34.4117647059%);fill:hsl(0, 0%, 34.4117647059%);}#diagram :root{--mermaid-font-family:"trebuchet ms",verdana,arial,sans-serif;}</style>
  <g>
    <marker id="diagram_flowchart-v2-pointEnd" class="marker flowchart-v2" viewBox="0 0 10 10" refX="5" refY="5" markerUnits="userSpaceOnUse" markerWidth="8" markerHeight="8" orient="auto">
      <path d="M 0 0 L 10 5 L 0 10 z" class="arrowMarkerPath" style="stroke-width: 1; stroke-dasharray: 1, 0;"></path>
    </marker>
    <marker id="diagram_flowchart-v2-pointStart" class="marker flowchart-v2" viewBox="0 0 10 10" refX="4.5" refY="5" markerUnits="userSpaceOnUse" markerWidth="8" markerHeight="8" orient="auto">
      <path d="M 0 5 L 10 10 L 10 0 z" class="arrowMarkerPath" style="stroke-width: 1; stroke-dasharray: 1, 0;"></path>
    </marker>
    <marker id="diagram_flowchart-v2-circleEnd" class="marker flowchart-v2" viewBox="0 0 10 10" refX="11" refY="5" markerUnits="userSpaceOnUse" markerWidth="11" markerHeight="11" orient="auto">
      <circle cx="5" cy="5" r="5" class="arrowMarkerPath" style="stroke-width: 1; stroke-dasharray: 1, 0;"></circle>
    </marker>
    <marker id="diagram_flowchart-v2-circleStart" class="marker flowchart-v2" viewBox="0 0 10 10" refX="-1" refY="5" markerUnits="userSpaceOnUse" markerWidth="11" markerHeight="11" orient="auto">
      <circle cx="5" cy="5" r="5" class="arrowMarkerPath" style="stroke-width: 1; stroke-dasharray: 1, 0;"></circle>
    </marker>
    <marker id="diagram_flowchart-v2-crossEnd" class="marker cross flowchart-v2" viewBox="0 0 11 11" refX="12" refY="5.2" markerUnits="userSpaceOnUse" markerWidth="11" markerHeight="11" orient="auto">
      <path d="M 1,1 l 9,9 M 10,1 l -9,9" class="arrowMarkerPath" style="stroke-width: 2; stroke-dasharray: 1, 0;"></path>
    </marker>
    <marker id="diagram_flowchart-v2-crossStart" class="marker cross flowchart-v2" viewBox="0 0 11 11" refX="-1" refY="5.2" markerUnits="userSpaceOnUse" markerWidth="11" markerHeight="11" orient="auto">
      <path d="M 1,1 l 9,9 M 10,1 l -9,9" class="arrowMarkerPath" style="stroke-width: 2; stroke-dasharray: 1, 0;"></path>
    </marker>
    <g class="root">
      <g class="clusters"></g>
      <g class="edgePaths">
        <path d="M92.352,62L92.352,66.167C92.352,70.333,92.352,78.667,92.352,86.333C92.352,94,92.352,101,92.352,104.5L92.352,108" id="L_F_F.1_0" class="edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link" style="" marker-end="url(#diagram_flowchart-v2-pointEnd)"></path>
      </g>
      <g class="edgeLabels">
        <g class="edgeLabel">
          <g class="label" transform="translate(0, 0)">
            <foreignObject width="0" height="0">
              <div style="display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;" xmlns="http://www.w3.org/1999/xhtml" class="labelBkg">
                <span class="edgeLabel"></span>
              </div>
            </foreignObject>
          </g>
        </g>
      </g>
      <g class="nodes">
        <g class="node default" id="flowchart-F-0" transform="translate(92.3515625, 35)">
          <rect class="basic label-container" style="" rx="5" ry="5" x="-42.75" y="-27" width="85.5" height="54"></rect>
          <g class="label" style="" transform="translate(-27.75, -12)">
            <rect></rect>
            <foreignObject width="55.5" height="24">
              <div style="display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;" xmlns="http://www.w3.org/1999/xhtml">
                <span class="nodeLabel">
                  <p>Finance</p>
                </span>
              </div>
            </foreignObject>
          </g>
        </g>
        <g class="node default" id="flowchart-F.1-1" transform="translate(92.3515625, 151)">
          <rect class="basic label-container" style="" x="-84.3515625" y="-39" width="168.703125" height="78"></rect>
          <g class="label" style="" transform="translate(-54.3515625, -24)">
            <rect></rect>
            <foreignObject width="108.703125" height="48">
              <div style="display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;" xmlns="http://www.w3.org/1999/xhtml">
                <span class="nodeLabel">
                  <!-- *** Validation fails here. *** -->
                  <p>F.1 Reuse Know<br>your customer</p>
                </span>
              </div>
            </foreignObject>
          </g>
        </g>
      </g>
    </g>
  </g>
</svg>
```

**Comment by sideshowbarker** [comment]
What exactly is the error that the checker emits?

When I test with that snippet in an HTML (text/html) document, the checker doesn’t report any errors about the `br` element from that `<p>F.1 Reuse Know<br>your customer</p>` markup.

Are instead saying that if you serve that document to the checker with an XML MIME type — or otherwise cause the checker to parse it with an XML parser rather than an HTML parser — then an error gets reported about that `<br>` markup?

If so, then it’s not even the checker which is reporting that error, but instead the XML parser — because in an XML document, markup with just a `<br>` start tag isn’t well-formed XML.

Markup for `br` elements can only have no end tag when it’s used in an HTML documents. The HTML spec doesn’t say that the end tag for it can be omitted when it’s in a XML document.

**Comment by KDean-Dolphin** [comment]
I haven't run it lately, so I'll have to check again. The problem ultimately is that the content of `foreignObject` inside SVG should conform to the content type for `foreignObject`, not XML, so SVG can't go through pure XML validation without running into this kind of problem.

**Comment by sideshowbarker** [comment]
> I haven't run it lately, so I'll have to check again.

👍

> The problem ultimately is that the content of `foreignObject` inside SVG should conform to the content type for `foreignObject`, not XML, so SVG can't go through pure XML validation without running into this kind of problem.

It remains unclear to me that exactly that means.

In particular, I don’t understand what _“pure XML validation”_ means. As far as the kind of validation that this checker performs, there is only one kind of validation, and it is agnostic about whether the document has been parsed with an XML parser vs having been parsed by an XML parser.

The validation is performed post-parsing, and the validation code which does it is unaware of how the input was parsed into a document.

And my understanding is that there isn’t any such thing as “XML validation” — instead there’s just XML _well-formedness_ — which is just about whether a document is _syntactically_ well-formed according to XML syntax requirements.

And the XML syntax requirements are such that any input with markup with just a `<br>` start tag but no end tag is not even an XML document — because it’s not syntactically well-formed XML. Such a document just fails to be parsed by any XML parser.

And to be clear, given some particular input stream, there’s no way to being parsing with it an XML parser and then switch mid-stream into, say, into instead parsing `foreignObject` content with an HTML parser.

Instead, the input must either be parsed entirely using an XML parser, or else parsed entirely with an HTML parser.

**Comment by KDean-Dolphin** [comment]
The problem is the ambiguity in the SVG specification. The text says:

> The <foreignObject> element includes elements from a different XML namespace. In the context of a browser, it is most likely (X)HTML.

The first sentence requires an XML namespace. The second implies that non-X HTML is permitted.

If non-X HTML is permitted, then assume the following (ignoring any missing namespace declarations):

```html
<html>
  <head>
    ...
  </head>
  <body>
    ...
    Some text with<br>a line break.
    ....
    <svg>
      ...
      <foreignObject>
        More text with<br>a line break.
      </foreignObject>
      ...
    </svg>
    ...
  </body>
</html>
```

If non-X HTML is permitted, validation should proceed roughly as follows:

```
Validate HTML until hitting the `svg` element.
  Validate SVG until hitting `foreignObject` element.
    Validate foreign object as HTML.
  Resume validating SVG.
Resume validating HTML.
```

Whether we like it or not (and I'm definitely on the "not" side), the reality is that XHTML doesn't have much market penetration, and tools that construct HTML inside the `foreignObject` tag use libraries, such as the JavaScript DOM library, that happily replace `<br/>` with `<br>`. Mermaid has a significant user base, and it generates non-X HTML code. Even the patch that the Mermaid people tried to put in to address this failed, because of the DOM library replacement issue.

**Comment by sideshowbarker** [comment]
Short answer: I’d suggest:

- raising an issue or PR to get MDN fixed
- raising a bug against Mermaid

…because as far as I can see, there actually isn’t any bug in the SVG spec related to this, and no bug in the checker/validator.

Longer asnwer:

> The problem is the ambiguity in the SVG specification.

Is the quotation you cite actually from the SVG spec or is it instead from the MDN article at https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Element/foreignObject? Because that MDN article is what the issue description links to.

What the spec itself at https://svgwg.org/svg2-draft/embedded.html#ForeignObjectElement says is this:

> The ‘foreignObject’ element allows for inclusion of elements in **a non-SVG namespace**.

…that is, nothing about XML or HTML (in any form).

> The text says:
> 
> > The element includes elements from a different XML namespace. In the context of a browser, it is most likely (X)HTML.
> 
> The first sentence requires an XML namespace.

Yeah, that seems to just be a bug in the MDN article, as far as I can see. For one thing, the indefinite article _“an”_ in _“an XML namespace”_ implies there’s somewhere more than one “XML namespace”. But there isn’t. There’s only one XML namespace, and that literally is just the string `http://www.w3.org/XML/1998/namespace`. See https://infra.spec.whatwg.org/#namespaces.

> The second implies that non-X HTML is permitted.

Yeah that second sentence should better just be completely removed from MDN. It adds nothing of value to the docs.

All the spec is actually saying about what `foreignObject` is that is can be content in some namespace other than
`http://www.w3.org/2000/svg`; for example, these namespaces:

- `http://www.w3.org/1999/xhtml` (the HTML namespace)
-  `http://www.w3.org/1998/Math/MathML` (the MathML namespace)

Regardless, it’s important to understand that all has nothing at all to do with XML vs HTML. That is, it has nothing at all to do with the _syntax_ of the markup source. It’s strictly just about differing namespaces.

> If non-X HTML is permitted

The SVG spec isn’t stating anything about whether “non-X HTML” is permitted or not.

Specifically, if what you mean by _“non-X HTML”_ is, HTML that‘s not well-formed XML, then the SVG spec anyway never (and _can’t_ really) defines new rules for how an XML parser will parse any of the content, nor how an HTML parser will. Instead, the rules for parsing XML are defined in the XML spec, and the rules for parsing HTML are defined in the HTML spec. And the SVG is not attempting to state an new parsing rules.

> then assume the following (ignoring any missing namespace declarations):
> ```html
> <html>
>   <head>
>     ...
>   </head>
>   <body>
>     ...
>     Some text with<br>a line break.
>     ....
>     <svg>
>       ...
>       <foreignObject>
>         More text with<br>a line break.
>       </foreignObject>
>       ...
>     </svg>
>     ...
>   </body>
> </html>
>```

If you run that document through an HTML parser, it will be parsed according to the parsing rules in the HTML spec — and will be parsed successfully into what you’d expect.

But if you run that document through an XML parser, it will be parsed according to the parsing rules in the XML spec — and will not be parsed successfully, because **it’s simply not well-formed XML**. Period.


> If non-X HTML is permitted, validation should proceed roughly as follows:
> 
> ```
> Validate HTML until hitting the `svg` element.
>   Validate SVG until hitting `foreignObject` element.
>     Validate foreign object as HTML.
>   Resume validating SVG.
> Resume validating HTML.
> ```

If you give that document as-is to an XML parser, it will never actually be validated at all. It will never reach validation — because it will fail parsing as XML. An XML parser will throw a fatal error for that document, and just stop.

> 
> Whether we like it or not (and I'm definitely on the "not" side), the reality is that XHTML doesn't have much market penetration, and tools that construct HTML inside the `foreignObject` tag use libraries, such as the JavaScript DOM library, that happily replace `<br/>` with `<br>`. Mermaid has a significant user base, and it generates non-X HTML code. Even the patch that the Mermaid people tried to put in to address this failed, because of the DOM library replacement issue.

In tools that actually work on a DOM there is no `<br/>` or `<br>` — instead there is just a `br` element, in memory in the DOM. In the DOM, that `br` element has no angle brackets around it. So DOM libraries don’t replace `<br/>` with `<br>`.

And so it’s unclear to me what Mermaid is doing and when. I have no idea what Mermaid even is or what it does. But based on what you’ve described, my closest guess is that either Mermaid is some thing that that _serializes_ a representation of the in-memory document in the DOM out a filesystem, or else that Mermaid is some thing that takes your markup source as input, and then per-processes it some way for some reason — and in doing that, it replaces instance of the markup `<br/>` with the markup `<br>`.

And then either way, my next guess is, that becomes a problem for you because you need to take that markup source which Mermaid has processed (mangled) and you need use it in some other system that uses an XML parser and which therefore assumes the document is well-formed XML.

Regardless, the only suggestion I can think of is: File a bug against Mermaid saying Please stop mangling `<br/>` into 
`<br>` (or whatever).

But again — as I said a the top of this comment — there’s otherwise nothing that needs changing here either in the SVG spec nor in the HTML checker/validator.

**Comment by sideshowbarker** [comment]
If https://github.com/validator/validator/issues/1816#issuecomment-3471564637 doesn’t resolve this issue for you, lemme know.

**Comment by sideshowbarker** [comment]
Going ahead and closing this because it does seem to be resolved as far as I can see.

**Comment by KDean-Dolphin** [comment]
Thanks for this. I'll take the documentation issue up with MDN. I agree with you that that's where the problem lies, but I'm stuck with the fact that [Mermaid](https://mermaid.js.org) has a significant installation base and generates non-X HTML in the output of the `foreignObject` tag. I've gone through various iterations with it, including setting the XML namespace inside the tag to force XHTML output, to no avail. This can stay closed.

**Comment by sideshowbarker** [comment]
Ok. Please do feel free to post further here if you reckon there's anything more I might be able to help with.
